### ISSUE TITLE
Output classification using LUT's

### ISSUE THREAD

**Issue by SampadChowdhury** [alanbacellar/DWN#1]
Hi,
First of all great work with the Differentiable LUT's. I am trying to grasp some concepts from this paper, but I've a couple of questions I would love to get clarified.
1. How is the output classification being done? The paper mentions about using LUT-2's instead of adders. Is it a complete replacement of adders in the circuit?
2. If we take example of mnist, does the LUT-2 tree structure ultimately comes down to just 10 LUT-2's for each class classification?
3. While the paper talks about using LUT-2's, the code present here uses merely the GroupSum technique(which in a sense is just adders in hardware). So, trying to understand if the Learnable Reduction thechnique is a part of  hardware implementation only? Or is it a part of the software training part as well?

I would really appreciate some inputs on these questions. 
Thankyou.

**Comment by alanbacellar** [comment]
Our approach replaces adders with a cascading pyramidal LUT structure (LUTs of any size, not necessarily LUT-2) specifically for the comparison with tiny classifiers. This is because, in very tiny circuits, adders become disproportionately large compared to the model itself. However, this substitution comes at the cost of accuracy. 

For all other experiments in the paper, we use popcount (adders) for reduction, as using LUTs instead leads to degraded accuracy.

Regarding the code, the pyramidal structure can be implemented using the `LUTLayer`. You can create sequential LUT layers with decreasing size and sequential mapping. For example, using LUT-2s:

```python
nn.Sequential(
    dwn.LUTLayer(32, 16, n=2, mapping='arange'),
    dwn.LUTLayer(16, 8, n=2, mapping='arange'),
    dwn.LUTLayer(8, 4, n=2, mapping='arange')
)
```
This example results in 4 output bits, allowing classification with up to 16 classes. For training, you apply binary cross-entropy on the individual bits.

**Comment by SampadChowdhury** [comment]
Thankyou for clarifying.
So, for datasets like MNIST/FashionMNIST(non TinyML datasets), the approach is still adders. But adders contribute a lot to the circuit area and power consumption. It would be really great if the learnable reduction technique could be used in datasets like MNIST/CIFAR. Is it known exactly like why using LUT's instead of adders leads to accuracy degradation?

**Comment by alanbacellar** [comment]
Yes. That's something we are currently working on. What we noticed is that accuracy goes down mostly due to gradient bottlenecks. Using summation in the end guarantees that all LUTs in the last layer receive gradients from the cross-entropy loss, but using a pyramidal structure can create a bottleneck depending on the LUT values.
